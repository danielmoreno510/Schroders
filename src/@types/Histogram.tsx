/* eslint-disable */
interface IHistogramProps {
  histogram: IHistogramData[];
  total: number;
  isLibrary: boolean;
  changeChart: () => void;
}

interface IHistogramData {
  name: string;
  label: string;
  value: number;
  size: number;
}

interface ISvgHistogramProps {
  data: IHistogramData[];
  total: number;
}
