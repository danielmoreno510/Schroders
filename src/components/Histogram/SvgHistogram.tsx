import React from 'react';
import {Svg, G, Line, Rect, Text} from 'react-native-svg';
import * as d3 from 'd3';
import {
  heightPercentageToDP,
  widthPercentageToDP,
} from '../../utils/responsiveUtils';
import {GRAPH_BAR_WIDTH, GRAPH_MARGIN, COLORS} from '../../constants';

const SvgHistogram: React.FC<ISvgHistogramProps> = ({data, total}) => {
  const SVGHeight = heightPercentageToDP('30%');
  const SVGWidth = '100%';
  const graphHeight = SVGHeight - 2 * GRAPH_MARGIN;
  const graphWidth = widthPercentageToDP('90%');

  const xDomain = data.map(item => item.label);
  const xRange = [0, graphWidth];
  const x = d3.scalePoint().domain(xDomain).range(xRange).padding(1);

  const maxValue = d3.max(data, (d: IHistogramData) => d.value)!;
  const topValue = Math.ceil(maxValue / data.length) * data.length;
  const yDomain = [0, topValue];
  const yRange = [0, graphHeight];
  const y = d3.scaleLinear().domain(yDomain).range(yRange);

  const middleValue = topValue / 2;

  return (
    <Svg width={SVGWidth} height={SVGHeight}>
      <G y={graphHeight + GRAPH_MARGIN}>
        <Text
          x={graphWidth - 4}
          textAnchor="end"
          y={y(topValue) * -1 - 10}
          fontSize={14}
          fill="black"
          fillOpacity={0.4}>
          {' '}
          {total}{' '}
        </Text>

        <Line
          x1="0"
          y1={y(topValue) * -1}
          x2={graphWidth}
          y2={y(topValue) * -1}
          stroke={COLORS.axis}
          strokeDasharray={[3, 3]}
          strokeWidth="0.5"
        />
        <Line
          x1="0"
          y1={y(middleValue) * -1}
          x2={graphWidth}
          y2={y(middleValue) * -1}
          stroke={COLORS.axis}
          strokeDasharray={[3, 3]}
          strokeWidth="0.5"
        />
        <Line
          x1="0"
          y1="2"
          x2={graphWidth}
          y2="2"
          stroke={COLORS.axis}
          strokeWidth="0.5"
        />

        {data.map(item => (
          <Rect
            key={item.label}
            x={x(item.label)! - GRAPH_BAR_WIDTH / 2}
            y={y(item.value) * -1}
            rx={2.5}
            width={GRAPH_BAR_WIDTH}
            height={y(item.value)}
            fill={COLORS.bars}
          />
        ))}

        {data.map(item => (
          <Svg key={item.label}>
            <Text fontSize="9" x={x(item.label)} y="12" textAnchor="middle">
              {' '}
              {item.label}{' '}
            </Text>
            <Text fontSize="7" x={x(item.label)} y="21" textAnchor="middle">
              {' '}
              {item.value}{' '}
            </Text>
          </Svg>
        ))}
      </G>
    </Svg>
  );
};

export default SvgHistogram;
