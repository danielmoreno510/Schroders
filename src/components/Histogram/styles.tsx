import {StyleSheet} from 'react-native';

export default StyleSheet.create({
  container: {
    paddingTop: 12,
  },

  titleContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingBottom: 24,
    justifyContent: 'space-between'
  },

  text: {
    fontSize: 18,
    textAlign: 'justify',
  },

  rowContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },

  month: {
    width: 80,
  },

  line: {
    marginTop: 4,
    marginBottom: 4,
    height: 24,
    justifyContent: 'center',
  },

  value: {
    textAlign: 'right',
    marginRight: 8,
    color: 'white',
    fontWeight: 'bold',
  },
});
