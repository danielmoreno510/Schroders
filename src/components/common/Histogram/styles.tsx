import {StyleSheet} from 'react-native';

export default StyleSheet.create({
  container: {
    paddingTop: 20,
  },

  text: {
    paddingBottom: 12,
    fontSize: 18,
    textAlign: 'justify',
  },

  rowContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },

  month: {
    width: 80,
  },

  line: {
    marginTop: 4,
    marginBottom: 4,
    height: 24,
    justifyContent: 'center',
  },

  value: {
    textAlign: 'right',
    marginRight: 8,
    color: 'white',
    fontWeight: 'bold',
  },
});
