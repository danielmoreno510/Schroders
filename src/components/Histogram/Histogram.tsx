import React from 'react';
import {Text, View} from 'react-native';

import styles from './styles';
import {widthPercentageToDP} from '../../utils/responsiveUtils';
import {INSET, primaryColor, secondaryColor} from '../../constants';
import SvgHistogram from './SvgHistogram';
import {NativeBaseProvider, Switch} from 'native-base';

const Histogram: React.FC<IHistogramProps> = ({
  histogram,
  total,
  isLibrary,
  changeChart,
}) => (
  <NativeBaseProvider initialWindowMetrics={INSET}>
    <View style={styles.container}>
      <View style={styles.titleContainer}>
        <Text style={styles.text}>Commits Histogram</Text>
        <Switch
          onTrackColor={primaryColor}
          isChecked={isLibrary}
          onToggle={changeChart}
        />
      </View>
      {isLibrary ? (
        <SvgHistogram data={histogram} total={total} />
      ) : (
        histogram.map(data => (
          <View key={data.name} style={styles.rowContainer}>
            <View style={styles.month}>
              <Text>{data.name}</Text>
            </View>
            {data.value > 0 ? (
              <View
                key={data.label}
                style={{
                  ...styles.line,
                  width: widthPercentageToDP(`${data.size}%`) - 80,
                  backgroundColor: secondaryColor,
                }}>
                <Text style={styles.value}>{data.value}</Text>
              </View>
            ) : (
              <View key={data.label} style={styles.line}>
                <Text style={{color: secondaryColor}}>{data.value}</Text>
              </View>
            )}
          </View>
        ))
      )}
    </View>
  </NativeBaseProvider>
);

export default Histogram;
