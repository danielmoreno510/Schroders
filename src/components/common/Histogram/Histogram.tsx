import React from 'react';
import {Text, View} from 'react-native';

import styles from './styles';
import {widthPercentageToDP} from '../../../utils/responsiveUtils';
import {secondaryColor} from '../../../constants';

const Histogram: React.FC<IHistogramProps> = ({histogram}) => (
  <View style={styles.container}>
    <Text style={styles.text}>Commits Histogram</Text>

    {histogram.map(data => (
      <View key={data.label} style={styles.rowContainer}>
        <View style={styles.month}>
          <Text>{data.label}</Text>
        </View>
        {data.value > 0 ? (
          <View
            key={data.label}
            style={{
              ...styles.line,
              width: widthPercentageToDP(`${data.size}%`) - 80,
              backgroundColor: secondaryColor,
            }}>
            <Text style={styles.value}>{data.value}</Text>
          </View>
        ) : (
          <View key={data.label} style={styles.line}>
            <Text style={{color: secondaryColor}}>{data.value}</Text>
          </View>
        )}
      </View>
    ))}
  </View>
);

export default Histogram;
