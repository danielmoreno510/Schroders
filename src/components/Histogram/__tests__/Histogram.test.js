import React from 'react';
import TestRenderer from 'react-test-renderer';

import Histogram from '../Histogram';

describe('Histogram component', () => {
  describe('GIVEN a set of valid props', () => {
    const testProps = {
      histogram: [
        {
          name: 'test',
          label: 'test',
          value: 1,
          size: 1,
        },
      ],
      total: 0,
      isLibrary: true,
      changeChart: jest.fn(),
    };

    describe('WHEN the component is rendered without props', () => {
      const testRenderer = TestRenderer.create(<Histogram {...testProps} />);

      it('THEN should display a regular histogram', () => {
        expect(testRenderer.toJSON()).toMatchSnapshot();
      });
    });
  });
});
