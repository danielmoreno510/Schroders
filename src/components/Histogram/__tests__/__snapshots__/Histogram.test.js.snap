// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Histogram component GIVEN a set of valid props WHEN the component is rendered without props THEN should display a regular histogram 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "paddingTop": 12,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingBottom": 24,
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 18,
            "textAlign": "justify",
          }
        }
      >
        Commits Histogram
      </Text>
      <RCTSwitch
        accessibilityRole="switch"
        activeThumbColor="#fafafa"
        colorScheme="primary"
        dataSet={{}}
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onTintColor="#f1b30a"
        style={
          [
            {
              "height": 31,
              "width": 51,
            },
            [
              {},
              {
                "backgroundColor": "#d4d4d4",
                "borderRadius": 16,
              },
            ],
          ]
        }
        thumbTintColor="#fafafa"
        tintColor="#d4d4d4"
        value={true}
      />
    </View>
    <RNSVGSvgView
      bbHeight="400"
      bbWidth="100%"
      focusable={false}
      height={400}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 400,
            "width": "100%",
          },
        ]
      }
      width="100%"
    >
      <RNSVGGroup
        fill={
          {
            "payload": 4278190080,
            "type": 0,
          }
        }
      >
        <RNSVGGroup
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          matrix={
            [
              1,
              0,
              -0,
              1,
              0,
              374,
            ]
          }
        >
          <RNSVGText
            content={null}
            dx={[]}
            dy={[]}
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            fillOpacity={0.4}
            font={
              {
                "fontSize": "14",
                "textAnchor": "end",
              }
            }
            propList={
              [
                "fill",
                "fillOpacity",
              ]
            }
            rotate={[]}
            x={
              [
                "671",
              ]
            }
            y={
              [
                "NaN",
              ]
            }
          >
            <RNSVGTSpan
              content=" "
              dx={[]}
              dy={[]}
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              font={{}}
              rotate={[]}
              x={[]}
              y={[]}
            />
            <RNSVGTSpan
              content="0"
              dx={[]}
              dy={[]}
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              font={{}}
              rotate={[]}
              x={[]}
              y={[]}
            />
            <RNSVGTSpan
              content=" "
              dx={[]}
              dy={[]}
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              font={{}}
              rotate={[]}
              x={[]}
              y={[]}
            />
          </RNSVGText>
          <RNSVGLine
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            propList={
              [
                "stroke",
                "strokeWidth",
                "strokeDasharray",
              ]
            }
            stroke={
              {
                "payload": 4294030090,
                "type": 0,
              }
            }
            strokeDasharray={
              [
                "3",
                "3",
              ]
            }
            strokeWidth="0.5"
            x1="0"
            x2="675"
            y1="NaN"
            y2="NaN"
          />
          <RNSVGLine
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            propList={
              [
                "stroke",
                "strokeWidth",
                "strokeDasharray",
              ]
            }
            stroke={
              {
                "payload": 4294030090,
                "type": 0,
              }
            }
            strokeDasharray={
              [
                "3",
                "3",
              ]
            }
            strokeWidth="0.5"
            x1="0"
            x2="675"
            y1="NaN"
            y2="NaN"
          />
          <RNSVGLine
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            propList={
              [
                "stroke",
                "strokeWidth",
              ]
            }
            stroke={
              {
                "payload": 4294030090,
                "type": 0,
              }
            }
            strokeWidth="0.5"
            x1="0"
            x2="675"
            y1="2"
            y2="2"
          />
          <RNSVGRect
            fill={
              {
                "payload": 4294030159,
                "type": 0,
              }
            }
            height="0"
            propList={
              [
                "fill",
              ]
            }
            rx="2.5"
            width="5"
            x="NaN"
            y="NaN"
          />
          <RNSVGSvgView
            bbHeight="100%"
            bbWidth="100%"
            focusable={false}
            style={
              [
                {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                {
                  "flex": 0,
                  "height": "100%",
                  "width": "100%",
                },
              ]
            }
          >
            <RNSVGGroup
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
            >
              <RNSVGText
                content={null}
                dx={[]}
                dy={[]}
                fill={
                  {
                    "payload": 4278190080,
                    "type": 0,
                  }
                }
                font={
                  {
                    "fontSize": "9",
                    "textAnchor": "middle",
                  }
                }
                rotate={[]}
                x={[]}
                y={
                  [
                    "12",
                  ]
                }
              >
                <RNSVGTSpan
                  content=" "
                  dx={[]}
                  dy={[]}
                  fill={
                    {
                      "payload": 4278190080,
                      "type": 0,
                    }
                  }
                  font={{}}
                  rotate={[]}
                  x={[]}
                  y={[]}
                />
                <RNSVGTSpan
                  content="test"
                  dx={[]}
                  dy={[]}
                  fill={
                    {
                      "payload": 4278190080,
                      "type": 0,
                    }
                  }
                  font={{}}
                  rotate={[]}
                  x={[]}
                  y={[]}
                />
                <RNSVGTSpan
                  content=" "
                  dx={[]}
                  dy={[]}
                  fill={
                    {
                      "payload": 4278190080,
                      "type": 0,
                    }
                  }
                  font={{}}
                  rotate={[]}
                  x={[]}
                  y={[]}
                />
              </RNSVGText>
              <RNSVGText
                content={null}
                dx={[]}
                dy={[]}
                fill={
                  {
                    "payload": 4278190080,
                    "type": 0,
                  }
                }
                font={
                  {
                    "fontSize": "7",
                    "textAnchor": "middle",
                  }
                }
                rotate={[]}
                x={[]}
                y={
                  [
                    "21",
                  ]
                }
              >
                <RNSVGTSpan
                  content=" "
                  dx={[]}
                  dy={[]}
                  fill={
                    {
                      "payload": 4278190080,
                      "type": 0,
                    }
                  }
                  font={{}}
                  rotate={[]}
                  x={[]}
                  y={[]}
                />
                <RNSVGTSpan
                  content="1"
                  dx={[]}
                  dy={[]}
                  fill={
                    {
                      "payload": 4278190080,
                      "type": 0,
                    }
                  }
                  font={{}}
                  rotate={[]}
                  x={[]}
                  y={[]}
                />
                <RNSVGTSpan
                  content=" "
                  dx={[]}
                  dy={[]}
                  fill={
                    {
                      "payload": 4278190080,
                      "type": 0,
                    }
                  }
                  font={{}}
                  rotate={[]}
                  x={[]}
                  y={[]}
                />
              </RNSVGText>
            </RNSVGGroup>
          </RNSVGSvgView>
        </RNSVGGroup>
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
</RNCSafeAreaProvider>
`;
